README.md
flex-twittersearch-widget
=========================

This widget is for monitoring Tweets within a specific geographic area. It was designed for disaster response efforts. You can combine its capabilities with other FlexViewer widgets. It includes searching Twitter within a geographic area, applying multiple filters against search results, automatic refresh, and clustering of results on map. 

####Main screen

![App](https://raw.github.com/esri/flex-twittersearch-widget/master/main_screen.png)

####Settings screen

![App](https://raw.github.com/esri/flex-twittersearch-widget/master/settings.png)

####Search results

![App](https://raw.github.com/esri/flex-twittersearch-widget/master/search_result.png)


##Installation

###Option 1

1) Copy twittersearch directory from the download into the widgets folder of your ArcGIS Viewer for Flex directory in FlashBuilder. The path should be \widgets\twittersearch. 
   - If you don't do this you will have lots of path errors when you try to run the widget.
   - If you don't have the the Viewer source code it can be downloaded here: 
     [https://github.com/ArcGIS/ArcGISViewerForFlex/tags](https://github.com/ArcGIS/ArcGISViewerForFlex/tags)
   - It's strongly recommended that you compile this widget with at least Adobe Flex SDK v4.6.
	 
2) Be sure to add twittersearch to config.xml and reference twittersearch/widgets/TwitterSearchWidget.swf, for example:

            <widget label="TwitterSearch"  x="850" y="100" preload="open" 
                icon="assets/images/i_twitter.png" 
                config="widgets/twittersearch/widgets/TwitterSearchWidget.xml"
                url="widgets/twittersearch/widgets/TwitterSearchWidget.swf"/>  

3) Configure twittersearch.xml to point to the location of your TwitterSearch ouath proxy. The project won't work without this. These have been included with your download in the /oauthproxy directory. 

4) Install PHP OAuth proxy. As of May 2013, Twitter will require all applications to use OAuth authentication. Versions of the TwitterSearch widget prioer to v4 used un-authenticated GET requests and will stop working as of May 2013. Here's the link to the [Twitter Search API doc](https://dev.twitter.com/docs/api/1.1/get/search/tweets).

**IMPORTANT: ONLY a PHP proxy is available as of v4.**

NOTE: Mac users. You will have to seperately install [Mcrypt](http://php.net/manual/en/book.mcrypt.php) libraries to test the proxy.

NOTE: Windows Users. I believe that the PHP for Windows installs the Mcrypt libraries automaticallyq: [http://windows.php.net/index.php](http://windows.php.net/index.php).

The widget won't work without this and you'll see an error when you try to run the widget. You do not have to recompile if you get the wrong proxy url. Simply change location in the xml, save the file,then restart the browser tab or window and the new proxy location should take effect. Sometimes, a user may have to flush their browser to ensure the changes took effect.
   
IMPORTANT: If you don't add a proxy directory, then when you go to run/debug the project it will throw an error. Once you set up the proxy this error will go away.

5) In FlashBuilder, under Project > Properties > Flex Modules: be sure you add a reference to the TwitterSearch module.This is an easy-to-forget step. But, if you don't do it, then the widget won't launch.

6) Copy as3corelib.swc from the root directory of the zip file into the lib directory of your Viewer project.

###Option #2: Deploy Immediately Option

The precompiled folder of the repo contains a precompiled version of the widget that's ready to be dropped onto a web server. You'll still need to configure twittersearch.xml, FlexViewer's config.xml, and set up the OAuth proxy.

##Configuring and Testing the OAuth Proxy (REQUIRED)

The proxy REQUIRES PHP and the Mcrypt library be installed. Mycrypt is typically installed when you use PHP for Windows. Mac users it might be easier to install PHP for Windows on a Windows Virtual Machine than go through the convolutions of installing Mycrypt on your Mac.

###Setting up Twitter Developer Account
- You WILL need a [Twitter Developer Account](https://dev.twitter.com/).
- Sign into your Twitter Developer Account.
- Hover over your Avatar (upper right hand side of page) and select "My Applications".
- Select Create New Application.
- The Application Website field needs the website of your organization. 
- The callback URL needs to be on a public website. You can still test locally even if you set this to a public website URL.
- Under the "Details" tab, copy the Consumer Key and Consumer Secret and paste those into the config.php file.
- Set the DOMAIN in config.php. Example for esri.com is ".esri.com".

###Configure the proxy

- Edit test.php by commenting out everything but the last two lines of code for generating the Random Key and IV. Run createKey.php to get your Key and IV. 
- Copy the Key and IV into config.php.
- Set the timezone in config.php.
- Run test.php. If everything is setup properly it should run with no errors.
- Sign into OAuth using sign_in.php
- Run the following command in your browser. If it fails to return values then something isn't set up correctly: [http://your_domain_name/oauthproxy/?cmd=search&geocode=34.26847173704432%2C-118.02937514220592%2C40mi&count=40&nocache=Thu%20Apr%204%2018](http://<your domain>/oauthproxy/?cmd=search&geocode=34.26847173704432%2C-118.02937514220592%2C40mi&count=40&nocache=Thu%20Apr%204%2018)
- Once this is configured you should be good to try and run the TwitterSearch widget inside FlexViewer.

If you simply can't get encryption to work then you can do the following (NOT RECOMMENDED and totally unsupported) steps. Most problems with encryption will stem from the Mcrypt library not being loaded.

- index.php comment out lines 15 thru 18 as well as 68, 69 and 70. Uncomment line 73. You might also have to comment out the require_once that points to Encrypt.php.
<code><pre>
	//$oauth_token = $encrypt->decrypt(base64_decode($access_token['oauth_token'])); echo "oauth ". (string)trim($oauth_token);
	//$oauth_token_secret = $encrypt->decrypt(base64_decode($access_token['oauth_token_secret'])); echo "\n\nsecret ". $oauth_token_secret."\n\n";
    //$connection = new TwitterOAuth(CONSUMER_KEY, CONSUMER_SECRET, (string)trim($oauth_token), (string)trim($oauth_token_secret));

	// Create a TwitterOauth object with consumer/user tokens.
    $connection = new TwitterOAuth(CONSUMER_KEY, CONSUMER_SECRET, $access_token['oauth_token'], $access_token['oauth_token_secret']);
</pre></code>    

- callback.php comment out the require_once pointing to Encrypt.php. Comment out lines 29, 30 and 31. Comment out lines 61 - 66.

##Minimum Requirements

- Adobe Flex 4.6. 

- The ArcGIS Flex API v3.x SDK which includes the swc library. This can be downloaded here: [http://resources.arcgis.com/en/communities/flex-api/](http://resources.arcgis.com/en/communities/flex-api/)

- Internet access to the ArcGIS Online sample servers.

- Web Server to deploy the application either Apache or IIS

- The ArcGIS Flex Viewer: [http://resources.arcgis.com/en/communities/flex-viewer/index.html](http://resources.arcgis.com/en/communities/flex-viewer/index.html)

- Flash Player 11.6 or greater

- oauthphp proxy. As of version 4 there is only one proxy and it runs on PHP.

- as3corelib.swc which is available in the root directory of the project.

- NOTE: The proxy included with this widget is not compatible as replacements for the Flex Viewer's own proxies. 

- Special NOTE: Twitter is subject to rate limiting. More details here [https://dev.twitter.com/docs/api/1.1/get/search/tweets](https://dev.twitter.com/docs/api/1.1/get/search/tweets) 

[](Esri Tags: FlexViewer Flex Widget Twitter OAuth)
[](Esri Language: ActionScript)

CHANGELOG.md
# flex-twittersearch-widget - Changelog

Changes in v4:

- Added support for Twitter OAuth which is required as of May 2013.
- Deprecated capabilities for non-authenticated search requests.
- Added a PHP-only OAuth proxy.
- Deprecated .NET and .JSP non-authenticated proxies. 
- Fixed bug that was preventing some geocoded tweets from displaying properly.
- Updated widget to work with ArcGIS API for Flex v3.2 and FlexViewer v3.2.

Changes in v3.1.1:
- Fixed minor bug that occured when reading Twitter geo property, caused message [Object, object] display

Changes in v3.1:
- Replaced XML tweet parser with a JSON tweet parser. Around late August Twitter modified their XML
  that caused parsing errors.
- Changed popup text color to white. Apparently this broke when upgrading to Flex 4.6
- Deprecated the need the xmlsyndication.swc.

Changes in v3:
- Upgraded widget to work with ArcGIS API for Flex v3
- Refactored many of the libraries to comply more with MVC pattern
- Commented out deprecated Graphic.autoMoveToTop property where appropriate as it has been deprecated.

Changes in v2.3.1:
- Upgraded widget to work with ArcGIs API for Flex 2.3.1.
- REMOVED ability for the widget to send tweets, now it can only search for tweets. All this code has been pulled. 
  The reason is related to changes that Twitter made to OAuth and the relative difficulty and time involved to
  implement those changes and ensure security. Also, very few people used or took advantage of this functionality. So,
  I apologize if you were using this code, but most likely your code wasn't working. 
  If you figure it out please send me a tweet using @agup.
- Tweaked the custom clustering algorythm so that now any tweet that does not have a geographic location is snapped
  to the middle of the search area. In prior versions the tweet simply did not show up on the map.
- Default timer now sets to 30 seconds instead of 1 minute. 

v2.1.2b known issues:
- Geolocated tweets pop up in an infowindow that corresponds to the actual tweet location rather than snapping to a
  cluster.
- Occasionally an avatar image will not display.
- Minor: When opening and closing the tweet login panel, it does not fully close.
- Very Minor: The oauth login verification landing page could be improved. It's very lightweight.

Changes in v2.1.2b:
- Upgraded to use production ArcGIS API for Flex v2.1, and the new production version of the Viewer v2.1.
- Organized the various files of the widget into logical directories.
- Updated many parts of the code to allow for easier testing and scalability.
- Removed alot of un-unused code that was taking up space.
- Fixed a bug that was preventing twitter geo-coded tweets from showing up on the map.
- Including an FXP in the download for those who simply want to import the project.
- Including an .NET/ashx version of the twitter proxy in the download.
- In the "settings" window you can now download a .CSV version of the tweets.

Changes in v2.0.1b:
- Upgraded widget to work with ArcGIS API for Flex v2, the new ArcGIS Viewer for Flex 2, Flex 4.x and Oauth 1.0a.
- Now includes an ashx proxy along with PHP and JSP.
- Removed the integrated Flash local store that held the twitter username and password. This is now handled directly
  through twitter because of the Oauth 1.0a requirements.

v1.0.1b Known issues:
- Occasionally a tweet will appear to successfully send from the widget, but the tweet didn't really get sent.
  Most often this is a problem with the twitter api. This is a documented problem where Twitter returns an HTTP 200
  status and they also send back an HTML page containing an error message. A future version of widget could be built to
  accommodate this, but we are hoping that twitter fixes their system to return a proper HTTP error code.
- Some Twitter avatars aren't displaying even though a valid image url was returned from Twitter.
- When scrolling through a popup window both a open hand cursor and pointer finger hand cursor display on top of each other.
- The reply symbol, which shows up when you hover over an avatar, isn't very visible against dark avatar images.
- There isn't any built-in help page or FAQ.
- There isn't an ashx proxy that works with the widget, there's only PHP and JSP.

Changes in v1.0.1b
 - Enabled the automatic zoom-in feature when user executes a new search.  
   It is controlled in TwitterSearchWidget.xml using the <zoomonresults> tag. Valid settings
   are true or false. Default is false which means don't zoom in.
 - Fixed: Search area circle was not projected correctly. Changed calculation to use 270 or 90 degrees.
 - Fixed: Widget not minimizing properly and still maintained GUI space. Moved messageBox into WidgetTemplate.
 - Fixed Login button enabled when browser cookies deleted and no username or password was visible.
 - Minor changes to OpenCloseHandler.as to tweak the UI handling.
 - Modified endpoint.php to lock it down so it only works with certain hostnames.
 - Changes to twittersearch.css.
 - Installation instructions rewritten and condensed into one section of the README.

oauthproxy/test.php
<?php
session_start();
require_once('config.php');
require_once('twitteroauth/Encrypt.php');
header('Content-Type: application/json');

$key = base64_decode(ENCRYPTION_KEY);
$iv = base64_decode(IV);
echo "Key: " . $key;
echo "\n\nIV: ". $iv;
$test = new Encrypt($key,$iv,DEFAULT_TIME_ZONE);
$temp = $test->encrypt("hello world");
echo "\n\nEncrypted Value: " . $temp;
echo "\n\nDecrypted Value: " . $test->decrypt($temp);
echo "\n\n---------\n\nGenerate Key (Base64): " . $test->get_RandomKey();
echo "\n\n---------\n\nGenerate IV : " . $test->get_IV();

?>

oauthproxy/callback.php
<?php

//Version 2.1 by AndyG 4/2013
//Changes
//- added OAuth Encrption

// Start session and load lib
session_start();
require_once('twitteroauth/twitteroauth.php');
require_once('twitteroauth/Encrypt.php');
require_once('config.php');

$content = null;    //for verification of credentials
$connection = null; //for getting access token

// check if cookie exists
if(isset($_COOKIE[OAUTH_COOKIE])){
    // redirect back to app
    if(isset($_SESSION['oauth_referrer'])){
        header('Location: '.$_SESSION['oauth_referrer']);
        exit;
    }
}
else{
    // if verifier set
    if(isset($_REQUEST['oauth_verifier'])){
	
		//Best practice is to encrypt the cookies or not use cookies
        $key = base64_decode(ENCRYPTION_KEY);
        $iv = base64_decode(IV);
        $encrypt = new Encrypt($key,$iv,DEFAULT_TIME_ZONE);


        // Create TwitteroAuth object with app key/secret and token key/secret from default phase
        $connection = new TwitterOAuth(CONSUMER_KEY, CONSUMER_SECRET, $_SESSION['oauth_token'], $_SESSION['oauth_token_secret']);

        // get access token from twitter
        try{
            $access_token = $connection->getAccessToken($_REQUEST['oauth_verifier']);
        }
        catch(Exception $e){
            header("HTTP/1.0 400 Error");
            echo "\n\nFailed retrieving access token: " .$e->getMessage();
            exit;
        }

        //Add a credentials validation request. Added v2.0 by AndyG
        try{
            $content = $connection->get('account/verify_credentials','');
        }
        catch(Exception $e){
            $error = $e->getMessage();
        }
        // save token
        $_SESSION['oauth_access_token'] = $access_token;
        // 1 year
        $cookie_life = time() + 31536000;

        if($content != null && $content->screen_name != ""){

			$token = base64_encode( $encrypt->encrypt($access_token['oauth_token']));
			$token_secret = base64_encode( $encrypt->encrypt($access_token['oauth_token_secret']));		
			
			//Update array with new encrypted values
			$access_token["oauth_token"] = $token;
			$access_token["oauth_token_secret"] = $token_secret;
			// echo "\n\n".var_dump($access_token); //for testing
			
			// set cookie
            setcookie(OAUTH_COOKIE, json_encode($access_token), $cookie_life, '/', OAUTH_COOKIE_DOMAIN);
            //header('Location: ./callback.php');
            echo "<html><head><title>Valid Verification</title><body bgcolor='#C0C0C0'>";
            echo "<style type='text/css'>body{font-family:sans-serif;}</style>";
            echo "<table width='100%'><tr bgcolor='#FFFFFF'><td>";
            echo "<a href='http://www.esri.com'><img src='edn.png' style='border-style:none' alt='ESRI Developer Network' /></a>";
            echo "</td></tr></table>";
            echo "<h2>Welcome:&nbsp;&nbsp;<img src='".$content->profile_image_url."'></img>&nbsp;&nbsp;&nbsp;@".$content->screen_name."</h2>";
            echo "<h4>You have successfully authenticated with Twitter. </h4>" ;
            echo "<h4>It is okay to close this page and return to the application.</h4>";
            echo "<script language=\"JavaScript\">\n";
            echo "if(window.opener && window.opener.getTokens){";
            echo "window.opener.getTokens(\"".$_SESSION['oauth_token'].",".$_SESSION['oauth_token_secret']."\");}";
            //You can also have the app automatically close the window via self.close, as shown below
            //echo "self.close();";
            echo "</script>";
            echo "</body></html>";
        }
        else{
            header("HTTP/1.0 400 Error");
            echo "\n\nFailed to validate credentials. ".$error;
            exit;
        }
        exit;
    }
    else{
       // redirect
        if(isset($_SESSION['oauth_referrer'])){
            header('Location: '.$_SESSION['oauth_referrer']);
        }
        else{
            header('Location: '.OAUTH_CALLBACK);
        }
        exit;
    }
}


oauthproxy/twitterimageproxy.php
<?php

//Version: 2.0.1b
//Creation Date: August 2010
//This proxy is designed to be used with the TwitterSearch FlexViewer widget
//IMPORTANT NOTICE:  This does NOT include any built-in security. It is strongly advised to lock down your proxy so that
//only authorized clients can use it, and that it only allows content to be referred to a specific destination.

$sURI = strtolower($_SERVER['REQUEST_METHOD']) == 'post' ? $_POST['uri'] : $_GET['uri'];

//$ch = curl_init("http://twitter.com/{$sURI}"); //use something like this to control where proxy can refer traffic
$ch = curl_init();

//Strip additional params off uri so we don't send them in final request stream
$shortURI = strtok($sURI,"&");

if ( strtolower($_SERVER['REQUEST_METHOD']) == 'post' && count($_POST) > 0 )
{
	curl_setopt($ch,CURLOPT_POST,TRUE);

	$aPost = array();
	foreach($_POST as $sIndex => $sValue)
	{
		$aPost[]="{$sIndex}=" . utf8_encode($sValue);
	}
	curl_setopt($ch, CURLOPT_POSTFIELDS,implode("&",$aPost));
}

$header = array(
     "Pragma: no-cache",
     "Cache-Control: max-age=0",
     "Expires: -1"
);

//Check for images. They require a binary stream and must be handled differently than text/html or xml
$ext = substr($sURI, -3);
switch ($ext) {
     case 'jpg': 
          $mime = 'image/jpeg';
          break; 
     case 'gif':
          $mime = 'image/gif'; 
          break; 
     case 'png':
          $mime = 'image/png';
          break; 
     case 'bmp':
          $mime = 'image/bmp';
          break;
     default: $mime = false; 
}

// Make sure the remote file is successfully opened before doing anything else.
// Treat image as binary otherwise you'll just get back garbage.
//To test if your proxy works with remote images try something like this: 
//http://<server path>/endpoint.php?uri=http://static.twitter.com/images/default_profile_normal.png
//If you get a 500 HTTP error make sure fopen() is enabled in php.ini
if ($mime && $file = @fopen($sURI, 'rb')) {
   
   header('Content-type: '.$mime);
   fpassthru($file);
   exit;   
}

curl_setopt($ch,CURLOPT_URL,$shortURI);
curl_setopt($ch,CURLOPT_HTTPHEADER, $header);
curl_setopt($ch,CURLOPT_RETURNTRANSFER, 1);
echo $result = curl_exec($ch);

if($result == false)
{
     error_log("endpoint.php - curl_exec error: ".curl_error($ch));
}

curl_close($ch);
?>

oauthproxy/twitteroauth/Encrypt.php
<?php
class Encrypt{
	/*
	* Basic encryption Class
	* Version: 1.0
	* Author: Andy Gup
	*/
    private $key;
	private $iv;
	
    function __construct($encryption_key,$iv,$time_zone) {
        $this->key = $encryption_key;
		$this->iv = $iv;
		date_default_timezone_set($time_zone);
    }

    public function encrypt($value){
        $final = null;
		if($value == null || $this->key == null || $this->iv == null)
		{
			header("HTTP/1.0 400 Error");
			echo "\n\n\nERROR: Null value detected: check your inputs";
			exit;
		}
		else
		{
			try
			{
				$final = mcrypt_encrypt(MCRYPT_RIJNDAEL_128, $this->key, $value, MCRYPT_MODE_CFB,$this->iv);
			}
			catch(Exception $e)
			{
				header("HTTP/1.0 400 Error");
				echo "\n\n\nERROR: Failed encryption. " .$e->getMessage();
				exit;
			}		
		}

        return  $final;
    }

    public function decrypt($value){
        return mcrypt_decrypt(MCRYPT_RIJNDAEL_128, $this->key, $value, MCRYPT_MODE_CFB,$this->iv);
    }

    public function get_RandomKey(){
        $result = openssl_random_pseudo_bytes(32, $secure);
        return base64_encode($result);
    }
	
	public function get_IV(){
		$size = mcrypt_get_iv_size(MCRYPT_CAST_256, MCRYPT_MODE_CFB);
		return base64_encode(mcrypt_create_iv($size, MCRYPT_RAND));
	}
}
?>

oauthproxy/twitteroauth/twitteroauth.php
<?php

/*
 * Abraham Williams (abraham@abrah.am) http://abrah.am
 *
 * The first PHP Library to support OAuth for Twitter's REST API.
 */

/* Load OAuth lib. You can find it at http://oauth.net */
require_once('OAuth.php');

/**
 * Twitter OAuth class
 */
class TwitterOAuth {
  /* Contains the last HTTP status code returned. */
  public $http_code;
  /* Contains the last API call. */
  public $url;
  /* Set up the API root URL. */
  public $host = "https://api.twitter.com/1.1/";
  /* Set timeout default. */
  public $timeout = 30;
  /* Set connect timeout. */
  public $connecttimeout = 30;
  /* Verify SSL Cert. */
  public $ssl_verifypeer = FALSE;
  /* Respons format. */
  public $format = 'json';
  /* Decode returned json data. */
  public $decode_json = TRUE;
  /* Contains the last HTTP headers returned. */
  public $http_info;
  /* Set the useragnet. */
  public $useragent = 'TwitterOAuth v0.2.0-beta2';
  /* Immediately retry the API call if the response was not successful. */
  //public $retry = TRUE;




  /**
   * Set API URLS
   */
  function accessTokenURL()  { return 'https://api.twitter.com/oauth/access_token'; }
  function authenticateURL() { return 'https://api.twitter.com/oauth/authenticate'; }
  function authorizeURL()    { return 'https://api.twitter.com/oauth/authorize'; }
  function requestTokenURL() { return 'https://api.twitter.com/oauth/request_token'; }
  function credentialsVerifyURL()  { return 'https://api.twitter.com/1/account/verify_credentials.json';}

  /**
   * Debug helpers
   */
  function lastStatusCode() { return $this->http_status; }
  function lastAPICall() { return $this->last_api_call; }

  /**
   * construct TwitterOAuth object
   */
  function __construct($consumer_key, $consumer_secret, $oauth_token = NULL, $oauth_token_secret = NULL) {
    $this->sha1_method = new OAuthSignatureMethod_HMAC_SHA1();
    $this->consumer = new OAuthConsumer($consumer_key, $consumer_secret);
    if (!empty($oauth_token) && !empty($oauth_token_secret)) {
      $this->token = new OAuthConsumer($oauth_token, $oauth_token_secret);
    } else {
      $this->token = NULL;
    }
  }


  /**
   * Get a request_token from Twitter
   *
   * @returns a key/value array containing oauth_token and oauth_token_secret
   */
  function getRequestToken($oauth_callback = NULL) {
    $parameters = array();
    if (!empty($oauth_callback)) {
      $parameters['oauth_callback'] = $oauth_callback;
    }
    $request = $this->oAuthRequest($this->requestTokenURL(), 'GET', $parameters);
    $token = OAuthUtil::parse_parameters($request);
    $this->token = new OAuthConsumer($token['oauth_token'], $token['oauth_token_secret']);
    return $token;
  }

  /**
   * Verify credentials
   *
   * @returns a string
   */
  function getVerifyCredentials() {
       $parameters = array();
       if (!empty($oauth_verifier)) {
            $parameters['oauth_verifier'] = $oauth_verifier;
       }
       $request = $this->oAuthRequest($this->credentialsVerifyURL(), 'GET', $parameters);
       $token = OAuthUtil::parse_parameters($request);
       //$this->token = new OAuthConsumer($token['oauth_token'], $token['oauth_token_secret']);
       return $token;
  }


  /**
   * Get the authorize URL
   *
   * @returns a string
   */
  function getAuthorizeURL($token, $sign_in_with_twitter = TRUE) {
    if (is_array($token)) {
      $token = $token['oauth_token'];
    }
    if (empty($sign_in_with_twitter)) {
      return $this->authorizeURL() . "?oauth_token={$token}";
    } else {
       return $this->authenticateURL() . "?oauth_token={$token}";
    }
  }

  /**
   * Exchange request token and secret for an access token and
   * secret, to sign API calls.
   *
   * @returns array("oauth_token" => "the-access-token",
   *                "oauth_token_secret" => "the-access-secret",
   *                "user_id" => "9436992",
   *                "screen_name" => "abraham")
   */
  function getAccessToken($oauth_verifier = FALSE) {
    $parameters = array();
    if (!empty($oauth_verifier)) {
      $parameters['oauth_verifier'] = $oauth_verifier;
    }
    $request = $this->oAuthRequest($this->accessTokenURL(), 'GET', $parameters);
    $token = OAuthUtil::parse_parameters($request);
    $this->token = new OAuthConsumer($token['oauth_token'], $token['oauth_token_secret']);
    return $token;
  }

  /**
   * One time exchange of username and password for access token and secret.
   *
   * @returns array("oauth_token" => "the-access-token",
   *                "oauth_token_secret" => "the-access-secret",
   *                "user_id" => "9436992",
   *                "screen_name" => "abraham",
   *                "x_auth_expires" => "0")
   */
  function getXAuthToken($username, $password) {
    $parameters = array();
    $parameters['x_auth_username'] = $username;
    $parameters['x_auth_password'] = $password;
    $parameters['x_auth_mode'] = 'client_auth';
    $request = $this->oAuthRequest($this->accessTokenURL(), 'POST', $parameters);
    $token = OAuthUtil::parse_parameters($request);
    $this->token = new OAuthConsumer($token['oauth_token'], $token['oauth_token_secret']);
    return $token;
  }

  /**
   * GET wrapper for oAuthRequest.
   */
  function get($url, $parameters = array()) { 		
    $response = $this->oAuthRequest($url, 'GET', $parameters);
    if ($this->format === 'json' && $this->decode_json) { 
      return json_decode($response);
    }
    return $response;
  }

  /**
   * POST wrapper for oAuthRequest.
   */
  function post($url, $parameters = array()) {
    $response = $this->oAuthRequest($url, 'POST', $parameters);
    if ($this->format === 'json' && $this->decode_json) {
      return json_decode($response);
    }
    return $response;
  }

  /**
   * DELETE wrapper for oAuthReqeust.
   */
  function delete($url, $parameters = array()) {
    $response = $this->oAuthRequest($url, 'DELETE', $parameters);
    if ($this->format === 'json' && $this->decode_json) {
      return json_decode($response);
    }
    return $response;
  }

  /**
   * Format and sign an OAuth / API request
   */
  function oAuthRequest($url, $method, $parameters) {
    if (strrpos($url, 'https://') !== 0 && strrpos($url, 'http://') !== 0) {
      $url = "{$this->host}{$url}.{$this->format}";
    }
    $request = OAuthRequest::from_consumer_and_token($this->consumer, $this->token, $method, $url, $parameters);
    $request->sign_request($this->sha1_method, $this->consumer, $this->token);
    switch ($method) {
    case 'GET':
      return $this->http($request->to_url(), 'GET');
    default:
      return $this->http($request->get_normalized_http_url(), $method, $request->to_postdata());
    }
  }

  /**
   * Make an HTTP request
   *
   * @return API results
   */
  function http($url, $method, $postfields = NULL) {
    $this->http_info = array();
    $ci = curl_init();
    /* Curl settings */
    curl_setopt($ci, CURLOPT_USERAGENT, $this->useragent);
    curl_setopt($ci, CURLOPT_CONNECTTIMEOUT, $this->connecttimeout);
    curl_setopt($ci, CURLOPT_TIMEOUT, $this->timeout);
    curl_setopt($ci, CURLOPT_RETURNTRANSFER, TRUE);
    curl_setopt($ci, CURLOPT_HTTPHEADER, array('Expect:'));
    curl_setopt($ci, CURLOPT_SSL_VERIFYPEER, $this->ssl_verifypeer);
    curl_setopt($ci, CURLOPT_HEADERFUNCTION, array($this, 'getHeader'));
    curl_setopt($ci, CURLOPT_HEADER, FALSE);

    switch ($method) {
      case 'POST':
        curl_setopt($ci, CURLOPT_POST, TRUE);
        if (!empty($postfields)) {
          curl_setopt($ci, CURLOPT_POSTFIELDS, $postfields);
        }
        break;
      case 'DELETE':
        curl_setopt($ci, CURLOPT_CUSTOMREQUEST, 'DELETE');
        if (!empty($postfields)) {
          $url = "{$url}?{$postfields}";
        }
    }

    curl_setopt($ci, CURLOPT_URL, $url);
    $response = curl_exec($ci);
    $this->http_code = curl_getinfo($ci, CURLINFO_HTTP_CODE);
    $this->http_info = array_merge($this->http_info, curl_getinfo($ci));
    $this->url = $url;
    curl_close ($ci);
    return $response;
  }

  /**
   * Get the header info to store.
   */
  function getHeader($ch, $header) {
    $i = strpos($header, ':');
    if (!empty($i)) {
      $key = str_replace('-', '_', strtolower(substr($header, 0, $i)));
      $value = trim(substr($header, $i + 2));
      $this->http_header[$key] = $value;
    }
    return strlen($header);
  }
}


oauthproxy/index.php
<?php

//Version 2.1 - Modifications by AndyG. 4/2013
//Changes
//- added encryption

// Load required lib files.
session_start();
require_once('twitteroauth/twitteroauth.php');
require_once('config.php');     
require_once('twitteroauth/Encrypt.php');
header('Content-Type: application/json');

$access_token = false;
$key = base64_decode(ENCRYPTION_KEY);
$iv = base64_decode(IV);
$encrypt = new Encrypt($key,$iv,DEFAULT_TIME_ZONE);

if(isset($_COOKIE[OAUTH_COOKIE])){
    // get access token from cookie

    $access_token = json_decode($_COOKIE[OAUTH_COOKIE], true);
    //Some systems may not decode json properly if it's escaped
    if($access_token == null){
        $access_token = json_decode(str_replace ('\"','"', $_COOKIE[OAUTH_COOKIE]), true);
        if($access_token == null){
            header("HTTP/1.0 400 Error");
            switch (json_last_error()) {
                case JSON_ERROR_DEPTH:
                    echo ' - Maximum stack depth exceeded';
                break;
                case JSON_ERROR_STATE_MISMATCH:
                    echo ' - Underflow or the modes mismatch';
                break;
                case JSON_ERROR_CTRL_CHAR:
                    echo ' - Unexpected control character found';
                break;
                case JSON_ERROR_SYNTAX:
                    echo ' - Syntax error, malformed JSON';
                break;
                case JSON_ERROR_UTF8:
                    echo ' - Malformed UTF-8 characters, possibly incorrectly encoded';
                break;
                default:
                    echo ' - Unknown error';
                break;
            }
            exit;
        }
    }

}

// if token exists
if (isset($access_token['oauth_token']) && isset($access_token['oauth_token_secret'])) {

	//**************************
	// FOR TESTING
	//	$token = base64_encode( $encrypt->encrypt($access_token['oauth_token']));
	//	$token_secret = base64_encode( $encrypt->encrypt($access_token['oauth_token_secret']));	
	//  $oauth_token = $encrypt->decrypt(base64_decode($token)); 
	//		echo "oauth ". (string)trim($oauth_token).",".$access_token['oauth_token'];
	//	$oauth_token_secret = $encrypt->decrypt(base64_decode($token_secret)); 
	//		echo "\n\nsecret ". $oauth_token_secret.", ".$access_token['oauth_token_secret']."\n\n";
	//**************************

    $oauth_token = $encrypt->decrypt(base64_decode($access_token['oauth_token'])); echo "oauth ". (string)trim($oauth_token);
	$oauth_token_secret = $encrypt->decrypt(base64_decode($access_token['oauth_token_secret'])); echo "\n\nsecret ". $oauth_token_secret."\n\n";
    $connection = new TwitterOAuth(CONSUMER_KEY, CONSUMER_SECRET, (string)trim($oauth_token), (string)trim($oauth_token_secret));
    // if invalid response	
	// Create a TwitterOauth object with consumer/user tokens.
    //$connection = new TwitterOAuth(CONSUMER_KEY, CONSUMER_SECRET, $access_token['oauth_token'], $access_token['oauth_token_secret']);
    
	// if invalid response
    if ($connection->http_code === 200 || $connection->http_code === 401) {
        $content = array('signedout'=>true);
    }
    else{
        // query params
        $params = array();
        if(isset($_REQUEST['q'])){
            $params['q'] = $_REQUEST['q'];
        }
        if(isset($_REQUEST['count'])){
            $params['count'] = $_REQUEST['count'];
        }
        if(isset($_REQUEST['result_type'])){
            $params['result_type'] = $_REQUEST['result_type'];
        }
        if(isset($_REQUEST['geocode'])){
            $params['geocode'] = $_REQUEST['geocode'];
        }
        if(isset($_REQUEST['max_id'])){
            $params['max_id'] = $_REQUEST['max_id'];
        }
        if(isset($_REQUEST['since_id'])){
            $params['since_id'] = $_REQUEST['since_id'];
        }
        if(isset($_REQUEST['include_entities'])){
            $params['include_entities'] = $_REQUEST['include_entities'];
        }
        if(isset($_REQUEST['lang'])){
            $params['lang'] = $_REQUEST['lang'];
        }
        if(isset($_REQUEST['until'])){
            $params['until'] = $_REQUEST['until'];
        }
        if(isset($_REQUEST['locale'])){
            $params['locale'] = $_REQUEST['locale'];
        }
        //Explicitly delete cookie - Added by AndyG. @ v2.0
        if(isset($_REQUEST['d'])){
            setcookie(OAUTH_COOKIE, '', 1, '/', OAUTH_COOKIE_DOMAIN);
            $content = array('signedout'=>true);
            exit;
        }
        //added in V2.0 by AndyG
        if(isset($_REQUEST['validate'])){
            $content = $connection->get('account/verify_credentials','');
            if($content != null){
                echo json_encode($content);
            }
            exit;
        }
        else{
        // call search
            $content = $connection->get('search/tweets', $params);
        }
        // if errors, signed out
        if (isset($content->errors) && count($content->errors)) {
            $content = array('signedout'=>true,'error'=>$content->errors); 
        }
    }
} else {
    // signed out
    $content = array('signedout'=>true);
}

// if callback set
if (isset($_REQUEST['callback'])) {
    echo $_REQUEST['callback'] . '(' . json_encode($content) . ');';
} else {
    echo json_encode($content);
}
